lib/at_host/at_host.c

```diff
@@ -112,76 +114,81 @@ static void cmd_send(struct k_work *work)
 
 static void uart_rx_handler(u8_t character)
 {
+	static bool cr_state; /* Whether last character received was a <CR> */
 	static bool inside_quotes;
-	static size_t cmd_len;
-	size_t pos;
+	static size_t at_cmd_len;
 
-	cmd_len += 1;
-	pos = cmd_len - 1;
-
-	/* Handle special characters. */
+	/* Handle control characters */
 	switch (character) {
 	case 0x08: /* Backspace. */
 		/* Fall through. */
 	case 0x7F: /* DEL character */
-		pos = pos ? pos - 1 : 0;
-		at_buf[pos] = 0;
-		cmd_len = cmd_len <= 1 ? 0 : cmd_len - 2;
-		break;
-	case '"':
-		inside_quotes = !inside_quotes;
-		 /* Fall through. */
-	default:
-		/* Detect AT command buffer overflow or zero length */
-		if (cmd_len > AT_MAX_CMD_LEN) {
-			LOG_ERR("Buffer overflow, dropping '%c'\n", character);
-			cmd_len = AT_MAX_CMD_LEN;
+		if (at_cmd_len > 0) {
+			at_cmd_len--;
+		}
+		return;
+	}
+
+	/*
+	 * Handle termination characters, if outside quotes.
+	 * The characters are never written to buffer unless inside quotes.
+	 */
+	if (!inside_quotes) {
+		switch (character) {
+		case '\0':
+			if (term_mode == MODE_NULL_TERM) {
+				goto send;
+			}
+			return;
+		case '\r':
+			if (term_mode == MODE_CR) {
+				goto send;
+			}
+			if (term_mode == MODE_CR_LF) {
+				cr_state = true;
+			}
 			return;
-		} else if (cmd_len < 1) {
-			LOG_ERR("Invalid AT command length: %d", cmd_len);
-			cmd_len = 0;
+		case '\n':
+			if (term_mode == MODE_LF) {
+				goto send;
+			}
+			if (term_mode == MODE_CR_LF && cr_state) {
+				goto send;
+			}
 			return;
 		}
-
-		at_buf[pos] = character;
-		break;
+		cr_state = false;
 	}
 
-	if (inside_quotes) {
+	/* Detect AT command buffer overflow, leaving space for null */
+	if (at_cmd_len + 1 > sizeof(at_buf) - 1) {
+		LOG_ERR("Buffer overflow, dropping '%c'\n", character);
 		return;
 	}
 
-	/* Check if the character marks line termination. */
-	switch (term_mode) {
-	case MODE_NULL_TERM:
-		/* Fall through. */
-	case MODE_CR:
-		if (character == termination[term_mode]) {
-			goto send;
-		}
-		break;
-	case MODE_LF:
-		if ((at_buf[pos - 1]) &&
-			character == termination[term_mode]) {
-			goto send;
-		}
-		break;
-	case MODE_CR_LF:
-		if ((at_buf[pos - 1] == '\r') && (character == '\n')) {
-			goto send;
-		}
-		break;
-	default:
-		LOG_ERR("Invalid termination mode: %d", term_mode);
-		break;
+	/* Write character to AT buffer */
+	at_buf[at_cmd_len] = character;
+	at_cmd_len++;
+
+	/* Handle special written character */
+	if (character == '"') {
+		inside_quotes = !inside_quotes;
 	}
 
 	return;
 send:
-	uart_irq_rx_disable(uart_dev);
-	k_work_submit(&cmd_send_work);
-	at_buf_len = cmd_len;
-	cmd_len = 0;
+	at_buf[at_cmd_len] = '\0'; /* Terminate the command string */
+
+	/* Reset UART handler state */
+	cr_state = false;
+	inside_quotes = false;
+	at_cmd_len = 0;
+
+	/* Send the command, if there is one to send */
+	if (at_buf[0]) {
```

Comment:
Do I understand the logic correctly, that there will be no `\r`, `\n` characters in the command string, and in  case of an empty line we'll end up with an empty string?

Reply:
Yes, the termination characters are not placed in the buffer. at_cmd will still hang if you send a command consisting only of whitespace chars, but that is a problem for a different PR, methinks.